armsrc/em4x50.c

```diff
@@ -0,0 +1,877 @@
+
+        if (i != length-1)
+            byte <<= 1;
+    }
+
+    return byte;
+}
+
+static uint8_t msb2lsb(uint8_t byte) {
+
+    // returns byte with bit positions in reverse order
+    
+    uint8_t tmp = 0;
+    
+    for (int i = 0; i < 8; i++)
+        tmp |= ((byte >> (7-i)) & 1) << i;
+    
+    return tmp;
+}
+
+static void msb2lsb_word(uint8_t *word) {
+    
+    // reorders given <word> according to EM4x50 datasheet (msb -> lsb)
+    
+    uint8_t buff[4];
+    
+    buff[0] = msb2lsb(word[3]);
+    buff[1] = msb2lsb(word[2]);
+    buff[2] = msb2lsb(word[1]);
+    buff[3] = msb2lsb(word[0]);
+
+    word[0] = buff[0];
+    word[1] = buff[1];
+    word[2] = buff[2];
+    word[3] = buff[3];
+}
+
+static void save_word(int pos, uint8_t bits[EM4X50_TAG_WORD]) {
+    
+    // split "raw" word into data, row and column parity bits and stop bit and
+    // save them in global tag structure
+    
+    uint8_t row_parity[4];
+    uint8_t col_parity[8];
+    
+    // data and row parities
+    for (int i = 0; i < 4; i++) {
+        tag.sectors[pos][i] = bits2byte(&bits[9*i],8);
+        row_parity[i] = bits[9*i+8];
+    }
+
+    tag.sectors[pos][4] = bits2byte(row_parity,4);
+
+    // column parities
+    for (int i = 0; i < 8; i++)
+        col_parity[i] = bits[36+i];
+
+    tag.sectors[pos][5] = bits2byte(col_parity,8);
+    
+    // stop bit
+    tag.sectors[pos][6] = bits[44];
+}
+
+static void wait_timer(int timer, uint32_t period) {
+
+    // do nothing for <period> using timer <timer>
+    
+    if (timer == FPGA_TIMER_0) {
+
+        AT91C_BASE_TC0->TC_CCR = AT91C_TC_SWTRG;
+        while (AT91C_BASE_TC0->TC_CV < period);
+
+    } else {
+
+        AT91C_BASE_TC1->TC_CCR = AT91C_TC_SWTRG;
+        while (AT91C_BASE_TC1->TC_CV < period);
+
+    }
+}
+
+static void em4x50_setup_read(void) {
+ 
+    FpgaDownloadAndGo(FPGA_BITSTREAM_LF);
+    
+    FpgaWriteConfWord(FPGA_MAJOR_MODE_LF_ADC | FPGA_LF_ADC_READER_FIELD);
+
+    // 50ms for the resonant antenna to settle.
+    SpinDelay(50);
+    // Now set up the SSC to get the ADC samples that are now streaming at us.
+    FpgaSetupSsc();
+    // start a 1.5ticks is 1us
+    StartTicks();
+ 
+    FpgaSendCommand(FPGA_CMD_SET_DIVISOR, LF_DIVISOR_125);
+
+    // Connect the A/D to the peak-detected low-frequency path.
+    SetAdcMuxFor(GPIO_MUXSEL_LOPKD);
+    
+    // Steal this pin from the SSP (SPI communication channel with fpga) and
+    // use it to control the modulation
+    AT91C_BASE_PIOA->PIO_PER = GPIO_SSC_DOUT;
+    AT91C_BASE_PIOA->PIO_OER = GPIO_SSC_DOUT;
+
+    // Disable modulation at default, which means enable the field
+    LOW(GPIO_SSC_DOUT);
+    
+    // Enable Peripheral Clock for
+    //   TIMER_CLOCK0, used to measure exact timing before answering
+    //   TIMER_CLOCK1, used to capture edges of the tag frames
+    AT91C_BASE_PMC->PMC_PCER |= (1 << AT91C_ID_TC0) | (1 << AT91C_ID_TC1);
+    AT91C_BASE_PIOA->PIO_BSR = GPIO_SSC_FRAME; // wofür ist das?
```

Comment:
Comment?   Please, use english.

Reply:
Oops - I forgot to delete it. What I wanted to say: "I have no idea what it means."

Reply:
deleted